crates/net/network/src/fetch/mod.rs

```diff
@@ -119,14 +127,27 @@ impl StateFetcher {
     }
 
     /// Returns the _next_ idle peer that's ready to accept a request,
-    /// prioritizing those with the lowest timeout/latency.
-    /// Once a peer has been yielded, it will be moved to the end of the map
-    fn next_peer(&mut self) -> Option<PeerId> {
-        self.peers
-            .iter()
-            .filter(|(_, peer)| peer.state.is_idle())
-            .min_by_key(|(_, peer)| peer.timeout())
-            .map(|(id, _)| *id)
+    /// prioritizing those with the lowest timeout/latency and those that recently responded with
+    /// adequate data.
+    fn next_best_peer(&mut self) -> Option<PeerId> {
+        let mut idle = self.peers.iter().filter(|(_, peer)| peer.state.is_idle());
+
+        let mut best_peer = idle.next()?;
+
+        for maybe_better in idle {
+            if best_peer.1.last_response_likely_bad && !maybe_better.1.last_response_likely_bad {
+                best_peer = maybe_better;
+                continue
+            }
+
+            if maybe_better.1.timeout() < best_peer.1.timeout() &&
+                !maybe_better.1.last_response_likely_bad
+            {
+                best_peer = maybe_better;
+            }
```

Comment:
nit: add small note on both these conds

```suggestion
            // replace best peer if our current best peer sent us a bad response last time
            if best_peer.1.last_response_likely_bad && !maybe_better.1.last_response_likely_bad {
                best_peer = maybe_better;
                continue
            }

            // replace best peer if this peer has better rtt
            if maybe_better.1.timeout() < best_peer.1.timeout() &&
                !maybe_better.1.last_response_likely_bad
            {
                best_peer = maybe_better;
            }
```

Reply:
> if our current best peer sent us a bad response last time

would replace with

> if our current best peer sent us a bad response in the past